### ISSUE TITLE
Websockets may drop (outgoing) messages

### ISSUE THREAD

**Issue by jxtps** [playframework/playframework#6246]
See: https://groups.google.com/d/msg/play-framework/GD8SzZk_Wo4/YL71YoKUBQAJ for additional background.

### Platform Etc
* Play Version: 2.5.4
* API: Scala (maybe also Java)
* Operating System: Any
* JDK: Any

### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.

1. Set up a websocket connection as described on [Handling WebSockets with Akka Streams and actors](https://www.playframework.com/documentation/2.5.x/ScalaWebSockets#Handling-WebSockets-with-Akka-Streams-and-actors)

2. Send >16 server -> client messages in your actor in one go, e.g. `for(i <- 0 until 20) out ! "Message " + i`

3. All messages should arrive at the client

### Actual Behavior

3. Some messages are dropped and never make it to the client. 

### Why this happens

`ActorFlow.actorRef()` uses a default buffer size of 16 and uses `OverflowStrategy.dropNew`, see [the source code](https://github.com/playframework/playframework/blob/d6c2673d91d85fd37de424951ee5ad9f4f4cce98/framework/src/play-streams/src/main/scala/play/api/libs/streams/ActorFlow.scala#L32). 

This means that if more than 16 messages are sent before they can be propagated out, some will be dropped. 

This breaks all reasonable expectations on the behavior of a websocket: it should not drop messages by default, and any fallback behavior should be explicitly called out and be required to be configured by the user. 

### Suggested Remedy

Either do not provide a default, i.e. require the user to explicitly set the bufferSize and make their own tradeoffs (together with a comment explaining the issue), 

OR 

Set the default to be `ActorMaterializer.settings.maxFixedBufferSize` so that a 128-slot fixed buffer is allocated, but if actual usage ever goes beyond 128 slots, then it automatically switches to a LinkedList-backed queue, never dropping messages. 

### Additional Concerns
It is not clear from the source code if there's a risk of incoming messages getting dropped - the buffer size appears to only affect outgoing messages.

**Comment by jxtps** [comment]
Tested the incoming messages with a simple for-loop on the client, sending 10k messages in one go and none were dropped, so incoming messages appear to be ok (have not verified by reviewing the source code - presumably the messages get sent to the receiving actor's inbox, and that one expands as needed).

**Comment by gmethvin** [comment]
@jxtps It seems like `OverflowStrategy.backpressure` is what people would expect as the default, right?

**Comment by jxtps** [comment]
Using `OverflowStrategy.backpressure` leads to `IllegalArgumentException: requirement failed: Backpressure overflowStrategy not supported`.

**Comment by schmitch** [comment]
@gmethvin actually I would suspect that it will honor the Mailbox size of my actor if I use actorRef and I think since 2.4 its unlimited and uses a linkedqueue.

**Comment by wpietri** [comment]
I wanted to second this. As somebody new to Akka and Play, the failure mode here is obscure. I spent a bunch of time looking for JS bugs, and then more time staring at Wireshark to discover that messages were getting dropped. I'd much prefer one of jxtps's solutions, or failing that making a noise when dropping messages (e.g., logging a warning or blowing up when the buffer is full).

**Comment by richdougherty** [comment]
I agree this is not ideal. The problem is that message passing to actors is inherently unreliable. To get reliability you generally need to layer higher-level flow control and error handling on top of your messages. It might be better to use an Akka Streams `Source` instead? This will handle backpressure, although you still won't get a notification when an error occurs, the `Source` will just be closed.

**Comment by wsargent** [comment]
@richdougherty I think that [`Sink.actorRefWithAck`](http://doc.akka.io/docs/akka/current/scala/stream/stream-integrations.html#Sink_actorRefWithAck) should be mentioned in the ScalaWebSockets page -- 2.5.1 has more options now, and it looks like @gmethvin has a backport going to Play 2.5.x.

**Comment by wpietri** [comment]
@richdougherty I agree that actors are in theory unreliable, and I'm sure actor experts have internalized that. But in practice, sending a few dozen lines over loopback via a websocket is quite reliable using other tools. Somebody copy-pasting the Play websocket example code could well be surprised here, and might easily conclude things were broken. I'd be happy with any solution that prevents that.

For my purposes it might be enough to change the docs to suggest whatever an expert would do in that case, with an explanation of what novices should be thinking about.

**Comment by wsargent** [comment]
If I remember correctly, this code was written before there was BroadcastHub available.  I think it should be possible to get something using https://github.com/playframework/play-scala-chatroom-example applied.

**Comment by wsargent** [comment]
Also see https://github.com/playframework/playframework/issues/6408

**Comment by ProChirathF** [comment]
Is there any workaround/pattern using ActorFlow with streaming data. I have tried Source.conflate to detect backpressure and to accumulate the output and also Source.actorRefWithAck to backpressure the WebSocket handler actor to no avail..

` ActorFlow.actorRef(MyWebSocketActor::props, actorSystem, materializer).conflate((a, b) -> a+b)
`

Is the recommendation of the framework to switch to streams for streaming data, this is not clear in the docs and looks really broken?

**Comment by ProChirathF** [comment]
I have implemented a Flow that acknowledges to user Websocket actor with WSAck once a message is emitted downstream, handling backpressure. User actor can use WSAck to implement a throttling mechanism. Below Java implementation is for Play2.7/Akka2.5, Play2.8/Akka2.6 should use Source.actorRefWithBackpressure .

```
public class BackpressureActorFlow {

	private final Logger logger = LoggerFactory.getLogger(BackpressureActorFlow.class);

	public static class WebSocketMessage {

		public long messageCount = 0;
		public String message = "";

		public WebSocketMessage(long messageCount, String message) {
			super();
			this.messageCount = messageCount;
			this.message = message;
		}

	}

	public static class WSAck {
	};

	public static Flow<String, String, NotUsed> actorRef(Function<ActorRef, Props> props, int bufferSize,
			ActorRefFactory factory, Materializer mat) {

		Source<WebSocketMessage, ActorRef> outerActor = Source.actorRefWithAck(new WSAck());
		Pair<ActorRef, Publisher<WebSocketMessage>> outerActoPublisherPair = outerActor
				.toMat(Sink.asPublisher(AsPublisher.WITHOUT_FANOUT), Keep.both()).run(mat);

		Sink<String, NotUsed> webSocketSink = Sink.actorRef(
				factory.actorOf(Props.create(SinkActor.class, props, outerActoPublisherPair.first())),
				new Success(CompletionStrategy.draining()));

		Source<String, NotUsed> webSocketSource = Source.fromPublisher(outerActoPublisherPair.second())
				.map(wsMessage -> {
					// logger.info("webSocketSource out {} {} ", wsMessage.messageCount);
					return wsMessage.message;
				}).addAttributes(Attributes.inputBuffer(1, bufferSize));

		Flow<String, String, NotUsed> flow = Flow.fromSinkAndSource(webSocketSink, webSocketSource);
		return flow;
	}

	static class SinkActor extends AbstractActor {
		private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

		public static Props props(final Function<ActorRef, Props> props, final ActorRef outerActor) {
			return Props.create(SinkActor.class, props, outerActor);
		}

		private final ActorRef flowActor;

		public SinkActor(Function<ActorRef, Props> props, ActorRef outerActor) {
			this.flowActor = getContext().watch(getContext().actorOf(props.apply(outerActor), "flowActor"));
		}

		@Override
		public Receive createReceive() {
			return receiveBuilder().match(Status.Success.class, x -> {
				flowActor.tell(PoisonPill.getInstance(), self());
			}).match(Status.Failure.class, x -> {
				flowActor.tell(PoisonPill.getInstance(), self());
			}).match(Terminated.class, x -> {
				getContext().stop(self());
			}).matchAny(o -> flowActor.tell(o, self())).build();
		}

		//TODO - supervisorStrategy() to return SupervisorStrategy.stop()
	}

}


```

**Comment by mkurz** [comment]
First, I think we should change `OverflowStrategy` to `fail` for `ActorFlow.actorRef()`. The interesting thing is the Java API already [does this](https://github.com/playframework/playframework/pull/5955/files#diff-9734b6c079e40531c22165528a767bc04a0911df750a8683b3fd9c35569ad5fdR70).

Second, I think we should provide an implementation using [`Source.actorRefWithBackpressure`](https://doc.akka.io/docs/akka/2.6.20/stream/operators/Source/actorRefWithBackpressure.html) and call it `ActorFlow.actorRefWithBackpressure()` and document that as the preferred one (probably deprecating the existing one).

Third, however, I think we should document `MergeHub` and `BroadcastHub` since they make a lot of sense when using websockets. Like already mentioned in this thread, it was not available back when this code was written.
There is a simple example already [here](https://github.com/playframework/play-samples/blob/971e00a68eae214a0710f27bc1758af05826e870/play-scala-chatroom-example/app/controllers/HomeController.scala#L32-L44).

One more thing to mention, like [mentioned in the original discussion](https://groups.google.com/g/play-framework/c/GD8SzZk_Wo4/m/YL71YoKUBQAJ), when using `ActorFlow.actorRef()`, one can set the buffer size to be `materializer.settings.maxFixedBufferSize`. That's because if the buffer size is greater or equal to MaxFixedBufferSize, [akka switches internally to a `BoundedBuffer`](https://github.com/akka/akka/blob/v2.6.20/akka-stream/src/main/scala/akka/stream/impl/Buffers.scala#L38-L39), which means it will, for the first 128 elements use an array and if more elements need to be buffered, [switches](https://github.com/akka/akka/blob/v2.6.20/akka-stream/src/main/scala/akka/stream/impl/Buffers.scala#L193) to [a `LinkedList`](https://github.com/akka/akka/blob/v2.6.20/akka-stream/src/main/scala/akka/stream/impl/Buffers.scala#L221). So in theory the buffer will never be to small, but of course there is a risk to run out of memory, which probably depends on your workload. Also see comments for [the config](https://github.com/akka/akka/blob/v2.6.20/akka-stream/src/main/resources/reference.conf#L56-L64),
